### ISSUE TITLE
Docker container runs processes as root/needs root file access

### ISSUE THREAD

**Issue by erdnapa** [signum-network/signum-node#559]
The signumnetwork/node container needs write access to run a phoenix update. This update process requires root file access permissions, preventing the processes in the container from running with user permissions only (via `-u $UID:$UID` for example).

To replicate:
`docker run -d --name signum --restart always -p 8121:8121 -p 8123:8123 -p 8125:8125 -v ~/docker/signum/conf:/conf/ -v ~/docker/signum/burst_db:/app/burst_db -u $UID:$UID signumnetwork/node:latest-h2` 

The above fails due to missing write access to update the phoenix wallet: 
`🛰 Updating Phoenix Wallet to current release...
mkdir: can't create directory '/app/tmp': Permission denied`

BRS itself/Java can run with user permissions (logging to file fails), but the update failure prevents its start.

Discord discussion for reference: https://discord.com/channels/813146842031849512/825303198939873280/899497275213111366 and following messages

Some background: https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b

@damccull

**Comment by gittrekt** [comment]
Due to docker limitations in general it's usually accepted practice to just run the application as a user to reduce risk.

I use mariadb and have modified my own docker container so I won't post everything here, but what I would do to achieve this would be to modify the Dockerfile.h2 and the start-node.sh to something like below 

# This is just a guess and may not work as I haven't tested. If you're going to try it, DO NOT use your existing DB directory, test with a different one.

### add this in the Dockerfile.h2 above the line _EXPOSE 8125 ...etc
`RUN_ useradd -rm -d /app -s /bin/bash -U signum`

### modify docker/start-node.sh to something like this
`chown -R signum:signum /app /db`
`chmod -R 766 /app /db`
`echo "🚀 Launching the node..."`
`exec runuser -u signum -- java -XX:MaxRAMPercentage=75.0 -jar $HOMEDIR/signum-node.jar --headless --config /app/conf`

This will run update-phoenix,sh and start-node.sh as root, update everything, then run signum as the user signum.

**Comment by damccull** [comment]
I did some experimenting with this and was having some permissions issues, but I didn't chown or chmod /db and /conf...I'll get back on it and try with those as well. Thanks for the suggestion.

**Comment by gittrekt** [comment]
> I did some experimenting with this and was having some permissions issues, but I didn't chown or chmod /db and /conf...I'll get back on it and try with those as well. Thanks for the suggestion.

I assume you'll still get permission errors on the logging (I do). I haven't tested if turning them off in the logging.properties suppresses them or not as I don't check my docker logs enough to care lol

**Comment by gittrekt** [comment]
> I did some experimenting with this and was having some permissions issues, but I didn't chown or chmod /db and /conf...I'll get back on it and try with those as well. Thanks for the suggestion.

Fixing logging issues can be resolved by changing the logging handler from:
`handlers = java.util.logging.FileHandler, java.util.logging.ConsoleHandler`
to just
`handlers = java.util.logging.ConsoleHandler`

No log files, but you'll get console logs and with chown you'll no longer have permission issues in general and run as user. 
Also if interested I would propose changing the start-node.sh to also download and update the node on each run instead of just phoenix. I can provide my start-node.sh if interested in reviewing for future releases.

**Comment by damccull** [comment]
I haven't had a chance to check this out yet but I'll check the docker stuff tonight or tomorrow.

As far as downloading the node, the reason I don't do that is because the container already builds it from source. I don't see a reason to not do that, so no reason to modify the script for that.

**Comment by gittrekt** [comment]
Yes that's a very valid point. I can argue both ways on it really as there are definitely cons. At first while using H2 my method caused corruption and I had to redownload a couple times so that's why I switched to mariadb. 

I find the benefits of using a release better for myself. I only have to restart the container and it's updated. I don't like to have to rebuild my container or cp a new version into the conf folder so I can copy it into the container to update. 

I've also had a recurring issue that some of the recent versions after building, it won't connect to any peers, but using a release in the same container connects fine. Downloading the release was a cheap and easy fix for me.

**Comment by damccull** [comment]
I suppose the container restart to update is just slightly easier than doing a `docker pull signumnetwork/node:latest-h2` or `docker run --pull always signumnetwork/node:latest-h2`, but in my opinion the whole point of a containerized software is to already be inside the container. The container doesn't build the software when you launch it. It builds it on github actions as part of the release cycle and then uploads the container with the fresh binary already in it to dockerhub.

The only reason it pulls phoenix on each start is because phoenix's release cadence is so much faster than the node's, it'd put phoenix several versions out of date by the next node release.

**Comment by gittrekt** [comment]
Yeah understandable, and I believe you're correct in that opinion really. It's entirely personal preference for me to use it more to keep my system clean from miscellaneous installs, and just let everything sit in a container. Also the issues I have with builds never connecting to peers when I do it the traditional way is another reason, but to be fair I haven't been bothered to investigate it as writing a bash script was just faster.
